File: shared/Services/MatchService.swift
import Combine
import Foundation
import os

/// Coordinates scoring engine + persistence. Owns the single active match workflow.
@MainActor
public final class MatchService: ObservableObject {
    @Published public private(set) var activeMatch: MatchState?
    @Published public private(set) var archivedMatches: [MatchState] = []
    @Published public private(set) var deletedMatchIDs: Set<UUID> = []
    @Published public private(set) var matchNotes: [UUID: String] = [:]
    @Published public private(set) var isRestored = false

    private let engine: ScoringEngine
    private let store: MatchStore
    private let logger = Logger(subsystem: "com.wristrally", category: "MatchService")
    private var syncHandler: ((MatchState?, [MatchState], Set<UUID>) -> Void)?

    public init(store: MatchStore, engine: ScoringEngine = ScoringEngine(), autoRestore: Bool = true) {
        self.store = store
        self.engine = engine
        if autoRestore {
            restore()
        }
    }

    /// Optional callback invoked after every persistence write (used by WatchConnectivity).
    public func onSyncNeeded(_ handler: @escaping (MatchState?, [MatchState], Set<UUID>) -> Void) {
        syncHandler = handler
    }

    public func restore() {
        do {
            deletedMatchIDs = try store.loadDeletedMatchIDs()
            matchNotes = try store.loadMatchNotes()
            activeMatch = try store.loadActiveMatch().map { engine.rehydrate($0) }
            archivedMatches = visibleArchive((try store.loadArchivedMatches()).map { engine.rehydrate($0) })
            logger.info("Restored active=\(self.activeMatch != nil) archive=\(self.archivedMatches.count)")
            finishRestore()
        } catch {
            logger.error("Restore failed: \(error.localizedDescription)")
            activeMatch = nil
            archivedMatches = []
            isRestored = true
        }
    }

    public func restore() async {
        do {
            let store = self.store
            let (active, archive, deleted, notes) = try await Task.detached(priority: .userInitiated) {
                let active = try store.loadActiveMatch()
                let archive = try store.loadArchivedMatches()
                let deleted = try store.loadDeletedMatchIDs()
                let notes = try store.loadMatchNotes()
                return (active, archive, deleted, notes)
            }.value
            deletedMatchIDs = deleted
            matchNotes = notes
            activeMatch = active.map { engine.rehydrate($0) }
            archivedMatches = visibleArchive(archive.map { engine.rehydrate($0) })
            logger.info("Restored active=\(self.activeMatch != nil) archive=\(self.archivedMatches.count)")
            finishRestore()
        } catch {
            logger.error("Restore failed: \(error.localizedDescription)")
            activeMatch = nil
            archivedMatches = []
            isRestored = true
        }
    }

    private func finishRestore() {
        isRestored = true
        expireInactiveMatchIfNeeded()
    }

    public func startMatch(settings: MatchSettings = .default) {
        guard activeMatch == nil else { return }
        let match = engine.startMatch(settings: settings)
        activeMatch = match
        persist()
        logger.info("Match started \(match.id.uuidString)")
    }

    public func awardPoint(to side: Side) {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.pointWon(side), to: match)
            activeMatch = match
            if match.status == .completed {
                finalizeActiveMatch()
            } else {
                persist()
                expireInactiveMatchIfNeeded()
            }
            logger.info("Point \(side.rawValue)")
        } catch {
            logger.error("Point failed: \(error.localizedDescription)")
        }
    }

    public func selectServer(_ side: Side) {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.selectServer(side), to: match)
            activeMatch = match
            persist()
            expireInactiveMatchIfNeeded()
            logger.info("Server selected \(side.rawValue)")
        } catch {
            logger.error("Select server failed: \(error.localizedDescription)")
        }
    }

    /// Asks who serves the set about to start, instead of carrying the rotation on.
    /// Offered at the changeover, where players often swap ends and reorder the serve.
    public func requestServerSelection() {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.requestServerSelection, to: match)
            activeMatch = match
            persist()
            expireInactiveMatchIfNeeded()
            logger.info("Server selection requested")
        } catch {
            logger.error("Server selection request failed: \(error.localizedDescription)")
        }
    }

    /// Ends the pre-match warm-up so the player can choose who serves.
    public func completeWarmUp() {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.completeWarmUp, to: match)
            activeMatch = match
            persist()
            logger.info("Warm-up completed")
        } catch {
            logger.error("Complete warm-up failed: \(error.localizedDescription)")
        }
    }

    public func undoLastPoint() {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.undo, to: match)
            activeMatch = match
            persist()
            expireInactiveMatchIfNeeded()
            logger.info("Undo")
        } catch {
            logger.error("Undo failed: \(error.localizedDescription)")
        }
    }

    public var canUndo: Bool {
        guard let match = activeMatch, match.status == .inProgress else { return false }
        return match.events.contains { $0.kind == .pointWon }
    }

    /// Switches how games are decided at 40-40 part-way through a match, for when the
    /// wrong format was picked at the start. Applies from the current point on; games
    /// already played keep the result they were scored with.
    public func setDeuceFormat(_ format: DeuceFormat) {
        guard var match = activeMatch, match.status == .inProgress else { return }
        guard match.settings.deuceFormat != format else { return }
        do {
            match = try engine.apply(.setDeuceFormat(format), to: match)
            activeMatch = match
            persist()
            logger.info("Deuce format changed to \(format.rawValue)")
        } catch {
            logger.error("Deuce format change failed: \(error.localizedDescription)")
        }
    }

    /// Updates in-match preference toggles without affecting scoring rules chosen at start.
    public func syncActiveMatchPreferences(
        usThemLabels: Bool,
        fixedServerPositions: Bool,
        askServeAtSetStart: Bool
    ) {
        guard var match = activeMatch, match.status == .inProgress else { return }
        match.settings.usThemLabels = usThemLabels
        match.settings.fixedServerPositions = fixedServerPositions
        match.settings.askServeAtSetStart = askServeAtSetStart
        activeMatch = match
        persist()
    }

    public func finishMatch() {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.finish, to: match)
            activeMatch = match
            finalizeActiveMatch()
            logger.info("Match finished")
        } catch {
            logger.error("Finish failed: \(error.localizedDescription)")
        }
    }

    /// Ends or discards an in-progress match that has exceeded the inactivity timeout.
    public func expireInactiveMatchIfNeeded(at date: Date = Date()) {
        guard let match = activeMatch, match.isInactive(at: date) else { return }
        if match.hasScoredPoints {
            endMatchEarly()
            logger.info("Match expired due to inactivity (ended early)")
        } else {
            discardMatch()
            logger.info("Match expired due to inactivity (discarded)")
        }
    }

    public func endMatchEarly() {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.endEarly, to: match)
            activeMatch = match
            finalizeActiveMatch()
            logger.info("Match ended early")
        } catch {
            logger.error("End early failed: \(error.localizedDescription)")
        }
    }

    public func discardMatch() {
        guard var match = activeMatch, match.status == .inProgress else { return }
        do {
            match = try engine.apply(.discard, to: match)
            // Discarded matches are not shown in history.
            activeMatch = nil
            try store.saveActiveMatch(nil)
            notifySync()
            logger.info("Match discarded")
        } catch {
            logger.error("Discard failed: \(error.localizedDescription)")
        }
    }

    /// Clears the active-match presentation after the user dismisses the completion screen.
    public func acknowledgeCompletedMatch() {
        guard let match = activeMatch, match.status.isTerminal, match.status != .discarded else {
            activeMatch = nil
            try? store.saveActiveMatch(nil)
            notifySync()
            return
        }
        // Ensure archived, then clear active.
        if !archivedMatches.contains(where: { $0.id == match.id }) {
            try? store.archiveMatch(match)
            archivedMatches = visibleArchive((try? store.loadArchivedMatches()) ?? archivedMatches)
        }
        activeMatch = nil
        try? store.saveActiveMatch(nil)
        notifySync()
    }

    /// Removes a match from history for good. The tombstone survives so a later Watch
    /// snapshot cannot resurrect it. The in-progress match is owned by the Watch and is
    /// never deleted here.
    public func deleteArchivedMatch(id: UUID) {
        guard id != activeMatch?.id else {
            logger.info("Refused to delete the active match \(id.uuidString)")
            return
        }
        deletedMatchIDs.insert(id)
        try? store.saveDeletedMatchIDs(deletedMatchIDs)
        try? store.deleteArchivedMatch(id: id)
        archivedMatches.removeAll { $0.id == id }
        if matchNotes.removeValue(forKey: id) != nil {
            try? store.saveMatchNotes(matchNotes)
        }
        notifySync()
        logger.info("Deleted archived match \(id.uuidString)")
    }

    /// Free text the user attached to a match on the phone. Notes live outside the
    /// synced `MatchState`, so they survive the Watch replacing the archive wholesale.
    public func note(for id: UUID) -> String {
        matchNotes[id] ?? ""
    }

    public func setNote(_ text: String, for id: UUID) {
        // A detail view saves its draft as it disappears, which includes being dismissed
        // by a delete — that must not resurrect the note it just pruned.
        guard !deletedMatchIDs.contains(id) else { return }
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let existing = matchNotes[id]
        if trimmed.isEmpty {
            guard existing != nil else { return }
            matchNotes.removeValue(forKey: id)
        } else {
            guard existing != trimmed else { return }
            matchNotes[id] = trimmed
        }
        do {
            try store.saveMatchNotes(matchNotes)
        } catch {
            logger.error("Saving note failed: \(error.localizedDescription)")
        }
    }

    /// Used by the phone when receiving Watch sync payloads.
    public func applyRemoteSnapshot(active: MatchState?, archive: [MatchState]) {
        activeMatch = active
        archivedMatches = visibleArchive(archive)
        try? store.saveActiveMatch(active)
        try? store.replaceArchive(archivedMatches)
    }

    /// Used by the Watch when the phone reports matches the user deleted from history.
    /// Archive only — the active match is never touched by a remote payload.
    public func applyRemoteDeletions(_ ids: Set<UUID>) {
        let incoming = Set(ids.filter { $0 != activeMatch?.id })
        guard !incoming.isEmpty else { return }
        deletedMatchIDs.formUnion(incoming)
        try? store.saveDeletedMatchIDs(deletedMatchIDs)
        let doomed = archivedMatches.filter { incoming.contains($0.id) }
        for match in doomed {
            try? store.deleteArchivedMatch(id: match.id)
        }
        archivedMatches.removeAll { incoming.contains($0.id) }
        logger.info("Applied \(incoming.count) remote deletion(s)")
    }

    private func visibleArchive(_ matches: [MatchState]) -> [MatchState] {
        matches
            .filter { $0.status != .discarded && !deletedMatchIDs.contains($0.id) }
            .sorted { $0.startedAt > $1.startedAt }
    }

    private func finalizeActiveMatch() {
        guard let match = activeMatch else { return }
        if match.status != .discarded {
            try? store.archiveMatch(match)
            archivedMatches = visibleArchive((try? store.loadArchivedMatches()) ?? [])
        }
        // Keep terminal match as active until user acknowledges Done (Watch completion screen).
        try? store.saveActiveMatch(match)
        notifySync()
    }

    private func persist() {
        do {
            try store.saveActiveMatch(activeMatch)
            notifySync()
        } catch {
            logger.error("Persist failed: \(error.localizedDescription)")
        }
    }

    private func notifySync() {
        syncHandler?(activeMatch, archivedMatches, deletedMatchIDs)
    }
}

File: android/domain/src/main/kotlin/com/wristrally/domain/MatchService.kt
package com.wristrally.domain

import java.time.Instant
import java.util.UUID
import java.util.logging.Logger

/** Coordinates scoring engine + persistence. Owns the single active match workflow. */
class MatchService(
    private val store: MatchStore,
    private val engine: ScoringEngine = ScoringEngine(),
    autoRestore: Boolean = true,
) {
    private val logger = Logger.getLogger("com.wristrally.MatchService")
    private val listeners = mutableListOf<() -> Unit>()

    var activeMatch: MatchState? = null
        private set(value) {
            field = value
            notifyListeners()
        }
    var archivedMatches: List<MatchState> = emptyList()
        private set
    var deletedMatchIDs: Set<UUID> = emptySet()
        private set
    var matchNotes: Map<UUID, String> = emptyMap()
        private set
    var isRestored: Boolean = false
        private set

    private var syncHandler: ((MatchState?, List<MatchState>, Set<UUID>) -> Unit)? = null

    init {
        if (autoRestore) restore()
    }

    fun addListener(listener: () -> Unit) {
        listeners += listener
    }

    fun removeListener(listener: () -> Unit) {
        listeners -= listener
    }

    fun onSyncNeeded(handler: (MatchState?, List<MatchState>, Set<UUID>) -> Unit) {
        syncHandler = handler
    }

    fun restore() {
        try {
            deletedMatchIDs = store.loadDeletedMatchIDs()
            matchNotes = store.loadMatchNotes()
            activeMatch = store.loadActiveMatch()?.let { engine.rehydrate(it) }
            archivedMatches = visibleArchive(store.loadArchivedMatches().map { engine.rehydrate(it) })
            logger.info("Restored active=${activeMatch != null} archive=${archivedMatches.size}")
            finishRestore()
        } catch (error: Exception) {
            logger.warning("Restore failed: ${error.message}")
            activeMatch = null
            archivedMatches = emptyList()
            isRestored = true
        }
    }

    private fun finishRestore() {
        isRestored = true
        expireInactiveMatchIfNeeded()
    }

    fun startMatch(settings: MatchSettings = MatchSettings()) {
        if (activeMatch != null) return
        val match = engine.startMatch(settings)
        activeMatch = match
        persist()
        logger.info("Match started ${match.id}")
    }

    fun awardPoint(to: Side) {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.PointWon(to), match)
            activeMatch = match
            if (match.status == MatchStatus.Completed) {
                finalizeActiveMatch()
            } else {
                persist()
                expireInactiveMatchIfNeeded()
            }
        } catch (error: Exception) {
            logger.warning("Point failed: ${error.message}")
        }
    }

    fun selectServer(side: Side) {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.SelectServer(side), match)
            activeMatch = match
            persist()
            expireInactiveMatchIfNeeded()
        } catch (error: Exception) {
            logger.warning("Select server failed: ${error.message}")
        }
    }

    fun requestServerSelection() {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.RequestServerSelection, match)
            activeMatch = match
            persist()
            expireInactiveMatchIfNeeded()
        } catch (error: Exception) {
            logger.warning("Server selection request failed: ${error.message}")
        }
    }

    fun completeWarmUp() {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.CompleteWarmUp, match)
            activeMatch = match
            persist()
        } catch (error: Exception) {
            logger.warning("Complete warm-up failed: ${error.message}")
        }
    }

    fun undoLastPoint() {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.Undo, match)
            activeMatch = match
            persist()
            expireInactiveMatchIfNeeded()
        } catch (error: Exception) {
            logger.warning("Undo failed: ${error.message}")
        }
    }

    val canUndo: Boolean
        get() {
            val match = activeMatch ?: return false
            if (match.status != MatchStatus.InProgress) return false
            return match.events.any { it.kind == MatchEventKind.PointWon }
        }

    fun setDeuceFormat(format: DeuceFormat) {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        if (match.settings.deuceFormat == format) return
        try {
            match = engine.apply(MatchAction.SetDeuceFormat(format), match)
            activeMatch = match
            persist()
        } catch (error: Exception) {
            logger.warning("Deuce format change failed: ${error.message}")
        }
    }

    fun syncActiveMatchPreferences(
        usThemLabels: Boolean,
        fixedServerPositions: Boolean,
        askServeAtSetStart: Boolean,
    ) {
        val match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        val updated = match.deepCopy()
        updated.settings = updated.settings.copy(
            usThemLabels = usThemLabels,
            fixedServerPositions = fixedServerPositions,
            askServeAtSetStart = askServeAtSetStart,
        )
        activeMatch = updated
        persist()
    }

    fun finishMatch() {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.Finish, match)
            activeMatch = match
            finalizeActiveMatch()
        } catch (error: Exception) {
            logger.warning("Finish failed: ${error.message}")
        }
    }

    fun expireInactiveMatchIfNeeded(at: Instant = Instant.now()) {
        val match = activeMatch ?: return
        if (!match.isInactive(at)) return
        if (match.hasScoredPoints) {
            endMatchEarly()
        } else {
            discardMatch()
        }
    }

    fun endMatchEarly() {
        var match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            match = engine.apply(MatchAction.EndEarly, match)
            activeMatch = match
            finalizeActiveMatch()
        } catch (error: Exception) {
            logger.warning("End early failed: ${error.message}")
        }
    }

    fun discardMatch() {
        val match = activeMatch ?: return
        if (match.status != MatchStatus.InProgress) return
        try {
            engine.apply(MatchAction.Discard, match)
            activeMatch = null
            store.saveActiveMatch(null)
            notifySync()
        } catch (error: Exception) {
            logger.warning("Discard failed: ${error.message}")
        }
    }

    fun acknowledgeCompletedMatch() {
        val match = activeMatch
        if (match == null || !match.status.isTerminal || match.status == MatchStatus.Discarded) {
            activeMatch = null
            runCatching { store.saveActiveMatch(null) }
            notifySync()
            return
        }
        if (archivedMatches.none { it.id == match.id }) {
            runCatching { store.archiveMatch(match) }
            archivedMatches = visibleArchive(
                runCatching { store.loadArchivedMatches() }.getOrDefault(archivedMatches),
            )
        }
        activeMatch = null
        runCatching { store.saveActiveMatch(null) }
        notifySync()
    }

    fun deleteArchivedMatch(id: UUID) {
        if (id == activeMatch?.id) return
        deletedMatchIDs = deletedMatchIDs + id
        runCatching { store.saveDeletedMatchIDs(deletedMatchIDs) }
        runCatching { store.deleteArchivedMatch(id) }
        archivedMatches = archivedMatches.filterNot { it.id == id }
        if (matchNotes.containsKey(id)) {
            matchNotes = matchNotes - id
            runCatching { store.saveMatchNotes(matchNotes) }
        }
        notifySync()
    }

    fun note(id: UUID): String = matchNotes[id].orEmpty()

    fun setNote(text: String, id: UUID) {
        if (deletedMatchIDs.contains(id)) return
        val trimmed = text.trim()
        val existing = matchNotes[id]
        val next = matchNotes.toMutableMap()
        if (trimmed.isEmpty()) {
            if (existing == null) return
            next.remove(id)
        } else {
            if (existing == trimmed) return
            next[id] = trimmed
        }
        matchNotes = next
        runCatching { store.saveMatchNotes(next) }
    }

    fun applyRemoteSnapshot(active: MatchState?, archive: List<MatchState>) {
        activeMatch = active
        archivedMatches = visibleArchive(archive)
        runCatching { store.saveActiveMatch(active) }
        runCatching { store.replaceArchive(archivedMatches) }
    }

    fun applyRemoteDeletions(ids: Set<UUID>) {
        val incoming = ids.filter { it != activeMatch?.id }.toSet()
        if (incoming.isEmpty()) return
        deletedMatchIDs = deletedMatchIDs + incoming
        runCatching { store.saveDeletedMatchIDs(deletedMatchIDs) }
        val doomed = archivedMatches.filter { incoming.contains(it.id) }
        for (match in doomed) {
            runCatching { store.deleteArchivedMatch(match.id) }
        }
        archivedMatches = archivedMatches.filterNot { incoming.contains(it.id) }
    }

    private fun visibleArchive(matches: List<MatchState>): List<MatchState> =
        matches
            .filter { it.status != MatchStatus.Discarded && !deletedMatchIDs.contains(it.id) }
            .sortedByDescending { it.startedAt }

    private fun finalizeActiveMatch() {
        val match = activeMatch ?: return
        if (match.status != MatchStatus.Discarded) {
            runCatching { store.archiveMatch(match) }
            archivedMatches = visibleArchive(
                runCatching { store.loadArchivedMatches() }.getOrDefault(emptyList()),
            )
        }
        runCatching { store.saveActiveMatch(match) }
        notifySync()
    }

    private fun persist() {
        try {
            store.saveActiveMatch(activeMatch)
            notifySync()
        } catch (error: Exception) {
            logger.warning("Persist failed: ${error.message}")
        }
    }

    private fun notifySync() {
        syncHandler?.invoke(activeMatch, archivedMatches, deletedMatchIDs)
    }

    private fun notifyListeners() {
        listeners.toList().forEach { it() }
    }
}
